config/versioning.py

```diff
@@ -0,0 +1,35 @@
+import os
+
+
+def get_git_sha():
+    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+    head_path = os.path.join(base_dir, ".git", "HEAD")
+
+    with open(head_path, "r") as fp:
+        head = fp.read().strip()
+
+    if head.startswith("ref: "):
+        head = head[5:]
+        revision_file = os.path.join(base_dir, ".git", *head.split("/"))
+    else:
+        return head
+
+    if not os.path.exists(revision_file):
+        # Check for Raven .git/packed-refs' file since a `git gc` may have run
+        # https://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery
+        packed_file = os.path.join(base_dir, ".git", "packed-refs")
+        if os.path.exists(packed_file):
+            with open(packed_file) as fh:
+                for line in fh:
+                    line = line.rstrip()
+                    if line and line[:1] not in ("#", "^"):
+                        try:
+                            revision, ref = line.split(" ", 1)
+                        except ValueError:
+                            continue
+                        if ref == head:
+                            return revision
```

Comment:
Would be nice to know and do - how to test these various points of returns in this code?

Reply:
I should probably document that this code is copied from:
https://github.com/getsentry/raven-python/blob/d7d14f61b7fb425bcb15512f659626648c494f98/raven/versioning.py#L11

Based on the advice from:
https://forum.sentry.io/t/fetch-git-sha-equivalent-in-the-unified-python-sdk/5521/2